src/Storage/TSStorageManager.m

```diff
@@ -248,42 +296,58 @@ - (NSData *)databasePassword
     [SAMKeychain setAccessibilityType:kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly];
 
     NSError *keyFetchError;
-    NSString *dbPassword =
-        [SAMKeychain passwordForService:keychainService account:keychainDBPassAccount error:&keyFetchError];
+    NSString *dbPassword = [SAMKeychain passwordForService:keychainService
+                                                   account:keychainDBPassAccount
+                                                     error:&keyFetchError];
 
     if (keyFetchError) {
-        switch (keyFetchError.code) {
-            case errSecItemNotFound:
-                dbPassword = [self createAndSetNewDatabasePassword];
-                break;
-            default:
-                DDLogError(@"%@ Getting DB password from keychain failed with error: %@", self.tag, keyFetchError);
-                [NSException raise:TSStorageManagerExceptionNameDatabasePasswordInaccessible
-                            format:@"Getting DB password from keychain failed with error: %@", keyFetchError];
-                break;
+        // Either this is a new install so there's no existing password to retrieve
+        // or the keychain has become corrupt.  Either way, we want to get back to a
+        // "known good state" and behave like a new install.
+        
+        BOOL shouldHavePassword = [[NSUserDefaults standardUserDefaults] objectForKey:TSStorageManagerHasDatabasePassword];
```

Comment:
Hmm, you're right.  I was trying to avoid filesystem access, but "less state = more surface area for bugs."